### PR TITLE
MBS-9592 Fix completely broken androidSdk build check

### PR DESCRIPTION
**Pull request by eugene-krivobokov**
It was broken due to a lack of integration tests.

### CODE REVIEW

File: subprojects/gradle/android/src/main/kotlin/com/avito/android/AndroidSdk.kt [hunk 1]
```diff
@@ -3,92 +3,73 @@ package com.avito.android
...
+    androidHome: File,
+    processRunner: ProcessRunner,
+    private val buildToolsVersion: String
+) : BaseAndroidSdk(androidHome, processRunner) {
 
     val aapt: Aapt
-        get() = Aapt.Impl(buildToolsPath(buildToolsVersion), processRunner)
+        get() = Aapt.Impl(ExistingDirectory.Impl(buildTools(buildToolsVersion)), processRunner)
 
-    val androidJar: File
```
Comment (eugene-krivobokov) [inline subprojects/gradle/android/src/main/kotlin/com/avito/android/AndroidSdk.kt:53]: not needed anymore

File: subprojects/gradle/android/src/main/kotlin/com/avito/android/AndroidSdk.kt [hunk 2]
```diff
@@ -3,92 +3,73 @@ package com.avito.android
...
-        get() {
-            val file = File(platformDir(compileSdkVersion), "android.jar")
-            require(file.exists()) {
-                sdkNotFoundMessage(file.path)
-            }
-            return file
-        }
+    companion object {
 
-    val platformSourceProperties: File
```
Comment (eugene-krivobokov) [inline subprojects/gradle/android/src/main/kotlin/com/avito/android/AndroidSdk.kt:62]: Don't need to know about this specific file in a common API.
Moved to a plugin.

File: subprojects/gradle/android/src/main/kotlin/com/avito/android/AndroidSdk.kt [hunk 3]
```diff
@@ -3,92 +3,73 @@ package com.avito.android
...
 
-class AndroidSdk(
-    private val project: Project,
-    private val processRunner: ProcessRunner
+open class BaseAndroidSdk(
+    val androidHome: File,
+    protected val processRunner: ProcessRunner
 ) {
-
-    private val buildToolsVersion: String
```
Comment (eugene-krivobokov) [inline subprojects/gradle/android/src/main/kotlin/com/avito/android/AndroidSdk.kt:17]: Non needed anymore

File: subprojects/gradle/android/src/main/kotlin/com/avito/android/AndroidSdk.kt [hunk 4]
```diff
@@ -3,92 +3,73 @@ package com.avito.android
...
-        get() {
-            require(project.isAndroid()) { "Trying to get android.buildToolsVersion object on non-android project: ${project.path}" }
-            return project.androidBaseExtension.buildToolsVersion
-        }
-
-    private val compileSdkVersion: Int
```
Comment (eugene-krivobokov) [inline subprojects/gradle/android/src/main/kotlin/com/avito/android/AndroidSdk.kt:23]: Moved this logic to a factory

File: subprojects/gradle/buildchecks/src/test/kotlin/com/avito/android/plugin/build_param_check/CheckAndroidSdkVersionTest.kt [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+import com.avito.test.gradle.TestResult
+import com.avito.test.gradle.dir
+import com.avito.test.gradle.file
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.io.TempDir
+import java.io.File
+import java.nio.file.Path
+
+class CheckAndroidSdkVersionTest {
```
Comment (eugene-krivobokov) [inline subprojects/gradle/buildchecks/src/test/kotlin/com/avito/android/plugin/build_param_check/CheckAndroidSdkVersionTest.kt:12]: Here are all corner cases

File: subprojects/gradle/android/src/main/kotlin/com/avito/android/AndroidSdk.kt [hunk 6]
```diff
@@ -3,92 +3,73 @@ package com.avito.android
...
 import java.util.Properties
 
 val Project.androidSdk: AndroidSdk
-    get() = AndroidSdk(this, ProcessRunner.Real())
+    get() = AndroidSdk.fromAndroidProject(this)
 
-class AndroidSdk(
-    private val project: Project,
-    private val processRunner: ProcessRunner
+open class BaseAndroidSdk(
```
Comment (eugene-krivobokov) [inline subprojects/gradle/android/src/main/kotlin/com/avito/android/AndroidSdk.kt:13]: Needed to separate two use cases: 
- SDK model based on Android module. Here we have BaseExtension with exact versions.
- Common Android SDK model. I've decided not to speculate here about any versions.

File: subprojects/gradle/buildchecks/src/main/kotlin/com/avito/android/plugin/build_param_check/CheckAndroidSdkVersionTask.kt [hunk 7]
```diff
@@ -41,19 +46,38 @@ abstract class CheckAndroidSdkVersionTask : DefaultTask() {
...
                     """
-            ${project.androidSdk.androidJar.path} has revision $localRevision but we use $expectedRevision in CI
-            It breaks build caching (https://issuetracker.google.com/issues/117789774)
-            Please update android-builder image: https://avito-tech.github.io/avito-android/docs/ci/containers/#android-builder-image
-            """
+                    You have a newer Android SDK Platform version.
+                    API level: ${compileSdkVersion.get()}, actual revision $localRevision, expected revision: $expectedRevision.
+                    It breaks build caching (https://issuetracker.google.com/issues/117789774).
+                    
+                    Please, update it in buildChecks config or in build environment.
```
Comment (sergioboi) [inline subprojects/gradle/buildchecks/src/main/kotlin/com/avito/android/plugin/build_param_check/CheckAndroidSdkVersionTask.kt:53]: Link to the `buildChecks` documentation?

**Review by sergioboi — APPROVED** [review]

File: subprojects/gradle/buildchecks/src/main/kotlin/com/avito/android/plugin/build_param_check/CheckAndroidSdkVersionTask.kt (hunk 7 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/buildchecks/src/main/kotlin/com/avito/android/plugin/build_param_check/CheckAndroidSdkVersionTask.kt:53]: Have it already in a final composite message
```
    You have a newer Android SDK Platform version.
    API level: 29, actual revision 6, expected revision: 5.
    It breaks build caching (https://issuetracker.google.com/issues/117789774).
    
    Please, update it in buildChecks config or in build environment.
    This check can be disabled by extension:
    buildChecks {
        androidSdk {
            enabled = false
        }
    }
    See https://avito-tech.github.io/avito-android/docs/projects/buildchecks/ (search 'androidSdk')
```
